### ISSUE TITLE
add "defer" attribute to JS

### ISSUE THREAD

**Issue by onlinebizsoft** [magento/magento2#33943]
### Summary (*)

Related to ticket https://github.com/magento/magento2/issues/8646

Magento offered Move JS to bottom but why dont you add an option to add "defer" attrbute to the JS file?
Having "defer" will make the browser rendering much faster.

### Proposed solution

I think adding "defer" works well with and without merging. It wont work with Bundling enabled only.

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users with no workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by onlinebizsoft** [comment]
@krzksz @mrtuvn @ihor-sviziev

**Comment by onlinebizsoft** [comment]
P/S there might be some issues with some JS using imperative notation with core modules but I think it is possible to work out.

**Comment by mrtuvn** [comment]
Default you can add defer via xml with attribute

**Comment by onlinebizsoft** [comment]
@mrtuvn I think you are talking about a different thing. I'm not talking about how requirejs load their components.
I mean for example when we use merge JS, we only have only single file included on the page -> we could make it "defer" (by a built-in Magento setting) and we need to make sure that all core modules comply with this this setting (I still see some core modules are using imperative notarion)

**Comment by mrtuvn** [comment]
Seem like another issue from my

**Comment by mrtuvn** [comment]
https://github.com/magento/magento2/issues/33686

**Comment by onlinebizsoft** [comment]
> #33686

I have seen but it is completely different indeed.

**Comment by mrtuvn** [comment]
Seem this one quite boat topic and not easy task in my opinion for resolve. After 5 years we still not solve it completely.  But still have another option for you. Hyva theme may be good choice for you. You will get complete new implements in frontend. I dont have work with that theme but i see they solution pretty work with current magento. Deffer js with alpine. Minimal js and with only 1 css. Minimal blocking render

**Comment by ihor-sviziev** [comment]
Hi @onlinebizsoft,
Could you send a PR with your suggestion? We'll review it

**Comment by onlinebizsoft** [comment]
> Hi @onlinebizsoft,
> Could you send a PR with your suggestion? We'll review it

I will do if we have an update to make this happening. Right now, I'm opening this for discussion from every one who land on the same issue :)
I also opened another ticket https://github.com/magento/magento2/issues/33946 -> it must be done first before this as well because Magento still have many render-blocking script

**Comment by ihor-sviziev** [comment]
I think we can't add 'defer' to the scripts, as it will break functionality, when some extension adds the `require` section as inline code.

**Comment by onlinebizsoft** [comment]
@ihor-sviziev if default Magento Luma do it well then all other extensions need to follow a standard and developers should understand what they need to do to comply with the setting

**Comment by ihor-sviziev** [comment]
Unfortunately, there are valid cases when such a case might be useful. For instance, adding some JS that should trigger almost immediately renders some piece that located on the first screen.

**Comment by onlinebizsoft** [comment]
@ihor-sviziev do you notice anything in the core now? When this is needed I think some small pure JS should be used and decided by developers.

**Comment by ihor-sviziev** [comment]
no, it's not in the core, but if you'll add it to the core - some extensions will be broken

**Comment by onlinebizsoft** [comment]
@ihor-sviziev it is up to the developers whom decide to enable it. So they need to make sure that all installed extensions are implemented in the best way.
And Magento make sure that all their core modules works fine, delivery the setting out of box.

In the official document, it recommends to avoid using imperative notation as well so it is good to have the defer setting which make the loading speed much higher

**Comment by mrtuvn** [comment]
Current i think magento team slowly try to remove unused parts. But it not easy and take efforts. Since from beginning it's bundle a lot of js. But most of it not used or unneccessary at some case
From my understand you want to inject js like this right ? Current we not able to do it without customise
![hyva-defer](https://user-images.githubusercontent.com/1908873/131513896-b5370445-b462-4c2a-8e8b-26687088c11f.png)

**Comment by engcom-Delta** [comment]
Hi @onlinebizsoft,
Thank you for raising an issue , Kindly provide issue description, steps to reproduce , actual results and expected results.

Regards,

**Comment by engcom-Delta** [comment]
We have noticed that this issue has not been updated for a period of more than 14 Days. 
Hence we assume that this issue is fixed now, so we are closing it. Please raise a fresh ticket or reopen this ticket if you need more assistance on this.
